x/virtualgroup/keeper/keeper.go

```diff
@@ -0,0 +1,684 @@
+	store := ctx.KVStore(k.storeKey)
+
+	seq := k.gvgFamilySequence.NextVal(store)
+	return seq
+}
+
+func (k Keeper) SetGVG(ctx sdk.Context, gvg *types.GlobalVirtualGroup) {
+	store := ctx.KVStore(k.storeKey)
+
+	bz := k.cdc.MustMarshal(gvg)
+	store.Set(types.GetGVGKey(gvg.Id), bz)
+}
+
+func (k Keeper) DeleteGVG(ctx sdk.Context, primarySpID, gvgID uint32) error {
+
+	store := ctx.KVStore(k.storeKey)
+
+	gvg, found := k.GetGVG(ctx, gvgID)
+	if !found {
+		return types.ErrGVGNotExist
+	}
+
+	if gvg.StoredSize != 0 {
+		return types.ErrGVGNotEmpty
+	}
+
+	gvgFamily, found := k.GetGVGFamily(ctx, primarySpID, gvg.FamilyId)
+	if !found {
+		panic("not found gvg family when delete gvg")
+	}
+
+	err := gvgFamily.RemoveGVG(gvg.Id)
+	if err == types.ErrGVGNotExist {
+		panic("gvg not found in gvg family when delete gvg")
+	}
+
+	for _, secondarySPID := range gvg.SecondarySpIds {
+		gvgStatisticsWithinSP := k.MustGetGVGStatisticsWithinSP(ctx, secondarySPID)
+		gvgStatisticsWithinSP.SecondaryCount--
+		k.SetGVGStatisticsWithSP(ctx, gvgStatisticsWithinSP)
+	}
+
+	store.Delete(types.GetGVGKey(gvg.Id))
+
+	k.SetGVGFamily(ctx, gvg.PrimarySpId, gvgFamily)
+	return nil
+}
+
+func (k Keeper) GetGVG(ctx sdk.Context, gvgID uint32) (*types.GlobalVirtualGroup, bool) {
+	store := ctx.KVStore(k.storeKey)
+
+	bz := store.Get(types.GetGVGKey(gvgID))
+	if bz == nil {
+		return nil, false
+	}
+
+	var gvg types.GlobalVirtualGroup
+	k.cdc.MustUnmarshal(bz, &gvg)
+	return &gvg, true
+}
+
+func (k Keeper) GetGVGByLVG(ctx sdk.Context, bucketID math.Uint, lvgID uint32) (*types.GlobalVirtualGroup, bool) {
+	lvg, found := k.GetLVG(ctx, bucketID, lvgID)
+	if !found {
+		return nil, false
+	}
+	gvg, found := k.GetGVG(ctx, lvg.GlobalVirtualGroupId)
+	if !found {
+		return nil, false
+	}
+	return gvg, true
+}
+
+func (k Keeper) SetLVG(ctx sdk.Context, lvg *types.LocalVirtualGroup) {
+	store := ctx.KVStore(k.storeKey)
+
+	bz := k.cdc.MustMarshal(lvg)
+	store.Set(types.GetLVGKey(lvg.BucketId, lvg.Id), bz)
+}
+
+func (k Keeper) GetLVG(ctx sdk.Context, bucketID math.Uint, lvgID uint32) (*types.LocalVirtualGroup, bool) {
+	store := ctx.KVStore(k.storeKey)
+
+	bz := store.Get(types.GetLVGKey(bucketID, lvgID))
+	if bz == nil {
+		return nil, false
+	}
+	var lvg types.LocalVirtualGroup
+	k.cdc.MustUnmarshal(bz, &lvg)
+	return &lvg, true
+}
+
+func (k Keeper) GetLVGs(ctx sdk.Context, bucketID math.Uint) []*types.LocalVirtualGroup {
+	lvgs := make([]*types.LocalVirtualGroup, 0)
+	store := ctx.KVStore(k.storeKey)
+	prefixStore := prefix.NewStore(store, types.GetLVGKey(bucketID, 0))
+	iterator := prefixStore.Iterator(nil, nil)
+	defer iterator.Close()
+
+	for ; iterator.Valid(); iterator.Next() {
+		var lvg types.LocalVirtualGroup
+		k.cdc.MustUnmarshal(iterator.Value(), &lvg)
+		lvgs = append(lvgs, &lvg)
+	}
+
+	return lvgs
+}
+
+func (k Keeper) GetGVGsBindingOnBucket(ctx sdk.Context, bucketID math.Uint) (*types.GlobalVirtualGroupsBindingOnBucket, bool) {
+	store := ctx.KVStore(k.storeKey)
```

Comment:
This is redundant metadata, as we have `GetLVGs`.  I think storage resources on-chain is more expensive than computation, hope we can use `GetLVGs` instead.

Reply:
But we can reduce up to 10 IO to 1 IO if we store this info. 

`GetLVGS` is very inefficient because it's a prefix iteration.